vmware-event-router/internal/processor/openfaas/openfaas_test.go

```diff
@@ -0,0 +1,324 @@
+// +build unit
+
+package openfaas
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"sync"
+	"testing"
+
+	"github.com/avast/retry-go"
+	ofsdk "github.com/openfaas-incubator/connector-sdk/types"
+	"github.com/pkg/errors"
+)
+
+// counter is concurrency-safe
+type counter struct {
+	sync.RWMutex
+	count int
+}
+
+func (c *counter) get() int {
+	c.RLock()
+	defer c.RUnlock()
+	return c.count
+}
+
+func (c *counter) increment() {
+	c.Lock()
+	defer c.Unlock()
+	c.count++
+}
+
+func Test_waitForAll(t *testing.T) {
+	// simulate numWait functions to wait for after their invocation
+	const numWait = 5
+	var c counter
+
+	ctx := context.Background()
+	ctxCancel, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// this func never fails
+	wfOk := func(ctx context.Context) error { return nil }
+
+	// this func fails based on a mod counter%numWait condition
+	wfFail := func(ctx context.Context) error {
+		if c.get()%numWait >= 2 {
+			return errors.New("failure occurred")
+		}
+		c.increment()
+		return nil
+	}
+
+	// this func fails with context cancelled err
+	wfFailCtx := func(ctx context.Context) error {
+		cancel()
+		return ctx.Err()
+	}
+
+	type args struct {
+		ctx   context.Context
+		waitN int
+		fn    waitFunc
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "All waitFn succeed",
+			args: args{
+				ctx:   ctx,
+				waitN: numWait,
+				fn:    wfOk,
+			},
+			wantErr: false,
+		},
+		{
+			name: "Partial failure",
+			args: args{
+				ctx:   ctx,
+				waitN: numWait,
+				fn:    wfFail,
+			},
+			wantErr: true,
+		},
+		{
+			name: "Context cancelled",
+			args: args{
+				ctx:   ctxCancel,
+				waitN: numWait,
+				fn:    wfFailCtx,
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := waitForAll(tt.args.ctx, tt.args.waitN, tt.args.fn); (err != nil) != tt.wantErr {
+				t.Errorf("waitForAll() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_waitForOne(t *testing.T) {
```

Comment:
These are really good tests. Maybe another one that would be good to add is one that ensures the timeout configuration works as intended.

Reply:
Added a test where a context has timed out and retry won't be performed.